libs/ledger/ledger/injections.py

```diff
@@ -109,27 +187,49 @@ def compare_metadata(cls, key, ours, theirs):
 class _WaveformGenerator:
     """Thin wrapper so that we can potentially parallelize this"""
 
-    gen: WaveformGenerator
+    waveform_approximant: str
     sample_rate: float
     waveform_duration: float
     coalescence_time: float
-
-    def shift_coalescence(self, waveform):
-        shift = int(self.coalescence_time * self.sample_rate)
-        return np.roll(waveform, shift, axis=-1)
+    minimum_frequency: float
+    reference_frequency: float
+
+    def shift_coalescence(self, waveform, t_final):
+        shift_time = t_final - (self.waveform_duration - self.coalescence_time)
+        shift_idx = int(shift_time * self.sample_rate)
+        return np.roll(waveform, shift_idx, axis=-1)
+
+    def align_waveforms(self, waveforms, t_final):
```

Comment:
So this will shift coalescence and pad if the requested waveform length is longer than the waveform duration?

Reply:
Yeah, `align_waveforms` pads and shifts if the generated waveform is shorter than the desired length, and shifts and crops if it's longer, which I think is the right way to arrange things.

For sure, I'll add some tests.

Reply:
Added some tests, let me know what you think